### PR TITLE
DI tests: identity

### LINKED ISSUE #1889: DI tests: identity
## Context

See #1682 

## Task

Implement each of the following test steps using the DI interface.
Refactor and extract common functionality as needed, and establish a pattern for the remaining test suites.
If needed, merge to dev in smaller batches to mitigate conflicts with other developments.

Along the way, consider better options for reducing the combinational explosion :P

### Steps

- [x] `check user sidechain storage before create`
- [x] `Invalid user shielding key`
- [x] `set user shielding key`
- [x] `check user shielding key from sidechain storage after setUserShieldingKey`
- [x] `check idgraph from sidechain storage before linking`
- [x] `link identities`
- [x] `check IDGraph after LinkIdentity`
- [x] `link invalid identities`
- [x] `link identities with wrong signature`
- [x] `deactivate identities`
- [x] `check IDGraph after deactivate identity`
- [x] `deactivate prime identity is disallowed`
- [x] `activate identity`
- [x] `check IDGraph after activate identity`
- [ ] `set error user shielding key`
- [ ] `exceeding IDGraph limit not allowed`

## Due Diligence

- [ ] Create issues for missing test cases from the original II test suite (marked with `TODO` comments in the code)

### PR DESCRIPTION
**Pull request by 0xverin**
### Context
resolves #1889

 Just completed the Substrate signer part; the EVM signer is currently in rapid progress. :p

### CODE REVIEW

File: tee-worker/ts-tests/integration-tests/common/transactions.ts [hunk 1]
```diff
@@ -217,3 +218,51 @@ export async function multiAccountTxSender(
...
+
+            const allBlockEvents = await shiftedApi.query.system.events();
+            const allExtrinsicEvents = allBlockEvents.filter(({ phase }) => phase.isApplyExtrinsic);
+
+            const matchingEvent = allExtrinsicEvents.find((eventRecord) => {
+                const eventData = eventRecord.event.data.toHuman();
+                /**
+                 * @FIXME I'd love a cleaner way to do this check :P
+                 */
+
```
Comment (arieljbirnbaum) [inline tee-worker/ts-tests/integration-tests/common/transactions.ts]: I don't think we have a better way to do this at the moment so I guess we should drop the `fixme` :(

File: tee-worker/ts-tests/integration-tests/common/transactions.ts [hunk 2]
```diff
@@ -217,3 +218,51 @@ export async function multiAccountTxSender(
...
+// for DI-test
+export const subscribeToEventsWithExtHash = async (
+    requestIdentifier: string,
+    context: IntegrationTestContext
+): Promise<FrameSystemEventRecord[]> => {
+    return new Promise<FrameSystemEventRecord[]>((resolve, reject) => {
+        let blocksToScan = 30;
+
+        const unsubscribe = context.api.rpc.chain.subscribeNewHeads(async (blockHeader) => {
```
Comment (arieljbirnbaum) [inline tee-worker/ts-tests/integration-tests/common/transactions.ts:237]: Maybe we should add a warning to our future selves here to _not_ `await unsubscribe` here :D

File: tee-worker/ts-tests/integration-tests/common/utils/assertion.ts [hunk 3]
```diff
@@ -5,19 +5,52 @@ import Ajv from 'ajv';
...
+) {
+    const failedType = context.api.events.identityManagement[eventType];
+    const isFailed = failedType.is.bind(failedType);
+    type EventLike = Parameters<typeof isFailed>[0];
+    const ievents: EventLike[] = events.map(({ event }) => event);
+    const failedEvent = ievents.filter(isFailed);
+    /* 
+      @fix Why this type don't work??????
+    */
+    const eventData = failedEvent[0].data[1] as CorePrimitivesErrorErrorDetail;
```
Comment (arieljbirnbaum) [inline tee-worker/ts-tests/integration-tests/common/utils/assertion.ts]: Please link to a followup issue? Maybe one of #1881 or #1917 could make sense :thinking:

File: tee-worker/ts-tests/integration-tests/common/utils/assertion.ts [hunk 4]
```diff
@@ -162,6 +197,19 @@ export async function assertIdentityActivated(
     console.log(colors.green('assertIdentityActivated complete'));
 }
 
+export async function assertWorkRpcReturnValue(callType: string, res: WorkerRpcReturnValue) {
```
Comment (arieljbirnbaum) [inline tee-worker/ts-tests/integration-tests/common/utils/assertion.ts]: `assertIsInSidechainBlock` ?

File: tee-worker/ts-tests/integration-tests/identity.test.ts [hunk 5]
```diff
@@ -135,6 +135,7 @@ describeLitentry('Test Identity', 0, (context) => {
 
         // Bob links:
         // - alice's substrate identity
+        // @FIXME: comment says alice; code says charlie. Which one is it supposed to be !?
         const aliceSubstrateIdentity = await buildIdentityHelper(
             u8aToHex(context.substrateWallet.charlie.addressRaw),
```
Comment (arieljbirnbaum) [inline tee-worker/ts-tests/integration-tests/identity.test.ts]: @Verin1005 I left this question there; do you know the answer?

Comment (0xverin) [reply tee-worker/ts-tests/integration-tests/identity.test.ts]: Naming error, it's clearly supposed to be 'charlie.' I'll make the correction here.

**Comment by 0xverin** [comment]
There is a testcase I haven't added, but I tested it locally:

1. Alice linked her identity without setting a shielding key before and then received `UserShieldingKeyNotFound` event data.

2. After that, when Alice tried to create shielding key, the work return status was invalid, and no parachain events appeared.

I'm not entirely sure if this is a bug or if I shouldn't be performing this operation like this? cc @kziemianek

**Comment by 0xverin** [comment]
Currently, [in all error cases example](https://github.com/litentry/litentry-parachain/blob/8ecf6944b4a792a8ebbebcaac75eed74f6eef70c/tee-worker/ts-tests/integration-tests/examples/direct-invocation/example.ts#L226), we receive an `isInvalid` status from trusted_call. However, we can still listen to events in the parachain. So added this [assertion](https://github.com/litentry/litentry-parachain/pull/1949/files#diff-09a700f38517a23b92015106656c7e4e7d79e3c9c3a649d9f84aa906f93dac29R121).

File: tee-worker/ts-tests/integration-tests/common/transactions.ts (hunk 1 above)
Comment (Verin1005) [reply tee-worker/ts-tests/integration-tests/common/transactions.ts]: removed.

File: tee-worker/ts-tests/integration-tests/common/utils/assertion.ts (hunk 3 above)
Comment (0xverin) [reply tee-worker/ts-tests/integration-tests/common/utils/assertion.ts]: linked.

File: tee-worker/ts-tests/integration-tests/common/utils/assertion.ts (hunk 4 above)
Comment (0xverin) [reply tee-worker/ts-tests/integration-tests/common/utils/assertion.ts]: changed.

File: tee-worker/ts-tests/integration-tests/identity.test.ts (hunk 5 above)
Comment (0xverin) [reply tee-worker/ts-tests/integration-tests/identity.test.ts]: fixed!

**Comment by kziemianek** [comment]
> There is a testcase I haven't added, but I tested it locally:
> 
> 1. Alice linked her identity without setting a shielding key before and then received `UserShieldingKeyNotFound` event data.
> 2. After that, when Alice tried to create shielding key, the work return status was invalid, and no parachain events appeared.
> 
> I'm not entirely sure if this is a bug or if I shouldn't be performing this operation like this? cc @kziemianek

1. Alice shouldn't be able to link identity without having shielding key set. If identity is linked then it looks like a bug.
2. Alice should be able to set shielding key at any time, if she can't then it looks like a bug too 🙃 

If you have ts-code reproducing this scenario - please share.

**Comment by Kailai-Wang** [comment]
> There is a testcase I haven't added, but I tested it locally:
> 
> 1. Alice linked her identity without setting a shielding key before and then received `UserShieldingKeyNotFound` event data.
> 2. After that, when Alice tried to create shielding key, the work return status was invalid, and no parachain events appeared.
> 
> I'm not entirely sure if this is a bug or if I shouldn't be performing this operation like this? cc @kziemianek

1 is expected
2 shouldn't happen

File: tee-worker/ts-tests/integration-tests/common/transactions.ts (hunk 2 above)
Comment (0xverin) [reply tee-worker/ts-tests/integration-tests/common/transactions.ts:237]: plz check that I've commented correctly, as I understand it

**Comment by 0xverin** [comment]
>Alice shouldn't be able to link identity without having shielding key set. If identity is linked then it looks like a bug.

I know it's expected 

>Alice should be able to set shielding key at any time, if she can't then it looks like a bug too.


Please change line 70,line 92 and line 100 charlie to alice, and then run it in locally and then:
- get `UserShieldingKeyNotFound` event which is expected
- setting user shielding key step will get invalid status

cc @kziemianek if you available.

**Comment by 0xverin** [comment]
~After that, when Alice tried to create shielding key, the work return status was invalid, and no parachain events appeared.~

Ohhhh, please disregard this comments. I retested it, and if I use 'alice' test the first step, I indeed need to update the nonce in the second step(I forgot). If the nonce is incorrect, the status will definitely be 'invalid'!" :PPPPPPP
cc @kziemianek @Kailai-Wang

**Comment by 0xverin** [comment]
I created a [follow-up comment](https://github.com/litentry/litentry-parachain/issues/1889#issuecomment-1651572091) to test the `evm-signer-identity-test`, and I will address it in another PR. Do you think it's better to continue adding test cases in the current(`identity-direct-invocation.test.ts`) file, or should I create a new file? I'm inclined to continue in the current file. cc @grumpygreenguy

**Comment by arieljbirnbaum** [comment]
@Verin1005 

> I created a [follow-up comment](https://github.com/litentry/litentry-parachain/issues/1889#issuecomment-1651572091) to test the `evm-signer-identity-test`, and I will address it in another PR. Do you think it's better to continue adding test cases in the current(`identity-direct-invocation.test.ts`) file, or should I create a new file? I'm inclined to continue in the current file. cc @grumpygreenguy

I think it'll be easier to have the EVM cases in a separate file for now; what I would suggest in any case is doing it in a separate PR so that this one can get merged :D

File: tee-worker/ts-tests/integration-tests/common/transactions.ts [hunk 6]
```diff
@@ -217,3 +218,54 @@ export async function multiAccountTxSender(
...
+// for DI-test
+
+export const subscribeToEventsWithExtHash = async (
+    requestIdentifier: string,
+    context: IntegrationTestContext
+): Promise<FrameSystemEventRecord[]> => {
+    return new Promise<FrameSystemEventRecord[]>((resolve, reject) => {
+        let blocksToScan = 30;
+        /* 
+        WARNING:The unsubscribe function is called inside the Promise callback, which is executed each time a new blockHeader is subscribed.
```
Comment (arieljbirnbaum) [inline tee-worker/ts-tests/integration-tests/common/transactions.ts]:
```suggestion
        WARNING:The unsubscribe function is called inside the Promise callback, which is executed each time a new blockHeader is received. 
```

File: tee-worker/ts-tests/integration-tests/common/transactions.ts (hunk 2 above)
Comment (arieljbirnbaum) [reply tee-worker/ts-tests/integration-tests/common/transactions.ts:237]: Looks OK; just suggested one amendment above :)

**Comment by 0xverin** [comment]
>I think it'll be easier to have the EVM cases in a separate file for now; what I would suggest in any case is doing it in a separate PR so that this one can get merged :D

 just worried that there are too many ts-testing processes in the CI.

**Comment by arieljbirnbaum** [comment]
> just worried that there are too many ts-testing processes in the CI.

Separate processes can be queued up if there's no capacity to run them tho; a longer process has a higher risk of just timing out :P

**Review by arieljbirnbaum — APPROVED** [review]
Let's ship it as soon as it's green :D

**Comment by Kailai-Wang** [comment]
Resolve conflicts and we are good to go

**Pull request merged**